### PR TITLE
feat: read-only database for querying the blockchain

### PR DESCRIPTION
**Pull request by feuGeneA**
Fixes https://github.com/NomicFoundation/rethnet/issues/49

### CODE REVIEW

File: crates/rethnet_evm/src/db/remote.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            client: RpcClient::new(url),
+            runtime: Builder::new_multi_thread()
+                .build()
+                .expect("failed to construct async runtime"),
+        }
+    }
+}
+
+impl DatabaseRef for RemoteDatabase {
+    type Error = anyhow::Error;
```
Comment (Wodann) [inline crates/rethnet_evm/src/db/remote.rs:31]: Please make this a dedicated error for `RemoteDatabase` using `thiserror`. I'm making the same changes for the other places I was originally using `anyhow::Error`

File: crates/rethnet_evm/src/db/remote.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
+use tokio::runtime::{Builder, Runtime};
+
+use rethnet_eth::remote::RpcClient;
+use rethnet_eth::{Address, B256, U256};
+
+use revm::{db::DatabaseRef, AccountInfo, Bytecode};
+
+#[derive(std::hash::Hash, std::cmp::Eq, std::cmp::PartialEq)]
+struct StorageLocation {
```
Comment (Wodann) [inline crates/rethnet_evm/src/db/remote.rs:9]: Is this used?

**Review by Wodann — CHANGES_REQUESTED** [review]

File: crates/rethnet_evm/src/db/remote.rs (hunk 2 above)
Comment (feuGeneA) [reply crates/rethnet_evm/src/db/remote.rs]: nope! must have been part of some previous draft! thanks! removed in the latest version of my commit.

File: crates/rethnet_evm/src/db/remote.rs (hunk 1 above)
Comment (feuGeneA) [reply crates/rethnet_evm/src/db/remote.rs]: I like it, thanks, done in the latest version of my commit.

File: crates/rethnet_evm/src/db/remote.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    #[error(transparent)]
+    RpcError(#[from] RpcClientError),
+
+    /// Some other error from an underlying dependency
+    #[error(transparent)]
+    OtherError(#[from] std::io::Error),
+}
+
+impl RemoteDatabase {
+    pub fn _new(url: &str) -> Self {
```
Comment (Wodann) [inline crates/rethnet_evm/src/db/remote.rs:27]:
```suggestion
    pub fn new(url: &str) -> Self {
```

Comment (Wodann) [reply crates/rethnet_evm/src/db/remote.rs:27]: You can avoid the warning by adding it to the public API of the crate

File: packages/hardhat-core/src/builtin-tasks/compile.ts [hunk 4]
```diff
@@ -94,7 +94,7 @@ function isConsoleLogError(error: any): boolean {
   return (
     error.type === "TypeError" &&
     typeof error.message === "string" &&
-    error.message.includes("log") &&
+    error.message.includes("log") === true &&
```
Comment (Wodann) [inline packages/hardhat-core/src/builtin-tasks/compile.ts:97]: Is this change needed?

File: packages/hardhat-etherscan/src/ABITypes.ts [hunk 5]
```diff
@@ -19,10 +19,10 @@ export function isABIArgumentLengthError(
 ): error is ABIArgumentLengthError {
   return (
     error.code === "INVALID_ARGUMENT" &&
-    error.count &&
+    Boolean(error.count) &&
```
Comment (Wodann) [inline packages/hardhat-etherscan/src/ABITypes.ts:22]: Are these changes needed?

**Review by Wodann — CHANGES_REQUESTED** [review]

**Review by Wodann — CHANGES_REQUESTED** [review]
There are still two open question about the changes in `*.ts` files

File: crates/rethnet_evm/src/db/remote.rs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                .build()
+                .expect("failed to construct async runtime"),
+        }
+    }
+
+    fn map_err(&self, err: RpcClientError) -> RemoteDatabaseError {
+        match err {
+            RpcClientError::SendError {
+                msg,
+                request_body: _,
```
[the remark this answers is not in the document]
Comment (feuGeneA) [reply crates/rethnet_evm/src/db/remote.rs]: FYI, this code isn't here anymore, because now I'm using  `thiserror`'s `transparent` functionality so that I don't have to re-code any field names here.

File: packages/hardhat-core/src/builtin-tasks/compile.ts (hunk 4 above)
Comment (feuGeneA) [reply packages/hardhat-core/src/builtin-tasks/compile.ts:97]: This is just Franco V fixing up linter errors in [this commit](https://github.com/NomicFoundation/hardhat/pull/3546/commits/ae0b12e92a76bf978bd804b983a0458750b0a5e3) as discussed [here](https://nomicfoundation.slack.com/archives/C03UN893ZT4/p1674118392149789?thread_ts=1674034616.578069&cid=C03UN893ZT4)

File: packages/hardhat-etherscan/src/ABITypes.ts (hunk 5 above)
Comment (feuGeneA) [reply packages/hardhat-etherscan/src/ABITypes.ts:22]: This is just Franco V fixing up linter errors in [this commit](https://github.com/NomicFoundation/hardhat/pull/3546/commits/ae0b12e92a76bf978bd804b983a0458750b0a5e3) as discussed [here](https://nomicfoundation.slack.com/archives/C03UN893ZT4/p1674118392149789?thread_ts=1674034616.578069&cid=C03UN893ZT4)

File: crates/rethnet_evm/src/db/remote.rs (hunk 3 above)
Comment (feuGeneA) [reply crates/rethnet_evm/src/db/remote.rs]: Addressed in https://github.com/NomicFoundation/hardhat/pull/3546/commits/a45659fcf4ce23da81ed14fbf3526117bb668b6a

File: packages/hardhat-core/src/builtin-tasks/compile.ts (hunk 4 above)
Comment (Wodann) [reply packages/hardhat-core/src/builtin-tasks/compile.ts:97]: Thanks!

**Review by Wodann — APPROVED** [review]
